treeppl/tree-inference-naive.tppl

```diff
@@ -0,0 +1,311 @@
+}
+
+
+/** 
+ * Stationary distribution of the transition matrix Q
+ *  Amounts to calculating the left eigenvector of the matrix
+ *  Hint:  https://brilliant.org/wiki/stationary-distributions/
+ */
+stationary_distribution = function(Q: InstantaneousRateMatrix) => Probability[4]
+{
+    // TODO implement stationary distribution
+}
+
+
+
+/// Tree Stuff ///
+
+/** Leaf 
+ * @param index the leaf number
+ * @param states the states of the leaf
+ * @return leaf
+ */
+leaf = function(index: BigInt, states: State[]) => Leaf
+{
+    return {
+	type: Leaf,
+	index: index,
+	age: 0.0,
+	states: states
+    }
+}
+
+
+/** Node */
+node = function(left: Tree | Leaf, right: Tree | Leaf, age: NonnegativeReal, states: State[]) => Tree
+{
+    return {
+	type: Node,
+	left: left,
+	right: right,
+	age: age,
+	states: states
+    }
+}
+
+/** Convert tree to Newick */
+top_to_newick = function(t: Tree)
+{
+    // TODO not implemented
+}
+
+/// Core model functions ///
+
+
+/** Recursive function to randomly build a tree */
+cluster = function(
+    Q: InstantaneousRateMatrix,
+    trees: Array<Tree>,
+    max_age : Age
+) => Array<Tree>
+{
+    // Check if we are done
+    if trees.length == 1
+        return trees
+
+    // Pick two random trees in the trees vector
+    v ~ Permutation(0:(trees.length - 1))
+    i = v[0]
+    j = v[1]
+
+    // Propose s (new node states) from arbitrary proposal distribution (here stationary distribution)
+    propose s ~ iid(
+	dist = Categorical(probabilities = stationary_distribution(Q), support = States),
+	n = trees[i].states.length
+    )
+    
+    // Propose node_age (new node age) from prior 
+    t ~ Exponential(rate = 20.0)
+    parental_age = max_age + t 
+
+    // Combine picked pair of trees into a new node
+    parental_node = node(left = trees[i], right = trees[j], age = parental_age, states = s)
+
+    // Observe states iid from continuous time Markov chain (CTMC)
+    // Note that trees[i].states can be observed if trees[i] is a leaf, otherwise it
+    // is a latent variable
+    // trees[i].states ~ iid( CTMC( parental_node.states, Q, parental_node.age - trees[i].age ) )
+    // trees[j].states ~ iid( CTMC( parental_node.states, Q, parental_node.age - trees[j].age ) )
+    // (comprehension style)
+    // These could probably be pre-computed
+    somehow because it is only 4 characters,
+    // and always the same time. So this is not the most efficient way to write it.
+    trees[i].states ~ Infer(
+	ctmc(
+	    parental_state = parental_node.states[k],
+	    evolution_time = parental_node.age - trees[i].age,
+	    Q = Q,
+	    P = transition_probabilities(Q)
+	)
+    ) where k in 0:(new_node.states.length - 1)
+    
+    trees[j].states ~ Infer(
+	ctmc(
+	    parental_state = parental_node.states[k],
+	    evolution_time = parental_node.age - trees[j].age,
+	    Q = Q,
+	    P = transition_probabilities(Q)
+	)
+    ) where k in 0:(new_node.states.length - 1)
```

Comment:
In principle, one could do it in this way, but then we need the Infer to be conditioned on the two end states, it is not sufficient to work your way back in time without conditioning on the starting state being the s[k] that we (potentially) sampled above. But if we need to sample change histories, it would be better not to do nested inference for each branch, but simply sample one change history for each particle. For computational efficiency, however, we should use the analytical transition probabilities that integrate out all change histories (see webppl script below). Furthermore, we should rely on delayed sampling to integrate out (sum over) the s[k] states when the number of states is small (say less than 64 states, which is the number of states for codon models).